pkg/resourcevalidator/info.go

```diff
@@ -0,0 +1,246 @@
+		}
+	}
+	pi.LastUpdateTime = time.Now().Format(time.RFC3339)
+}
+
+func (pi *peeringInfo) subtractResources(resources v1.ResourceList) {
+	for key, val := range resources {
+		if prevFree, ok := pi.FreePeeringQuota[key]; ok {
+			prevFree.Sub(val)
+			pi.FreePeeringQuota[key] = prevFree
+		} else {
+			pi.FreePeeringQuota[key] = val.DeepCopy()
+		}
+	}
+	pi.LastUpdateTime = time.Now().Format(time.RFC3339)
+}
+
+func createPeeringInfo(clusterID, clusterName string, resources v1.ResourceList) *peeringInfo {
+	return &peeringInfo{
+		clusterID:    clusterID,
+		clusterName:  clusterName,
+		SPList:       map[string]ShadowPodDescription{},
+		PeeringQuota: resources.DeepCopy(),
+		//UsedPeeringQuota: generateQuotaPattern(resources),
+		FreePeeringQuota: resources.DeepCopy(),
+		piMutex:          sync.RWMutex{},
+		LastUpdateTime:   time.Now().Format(time.RFC3339),
+	}
+}
+
+func (pi *peeringInfo) getClusterID() string {
+	return pi.clusterID
+}
+
+func (pi *peeringInfo) getClusterName() string {
+	return pi.clusterName
+}
+
+func (pi *peeringInfo) updatePeeringQuota(resources v1.ResourceList) {
+	webhookrefreshlog.Info(quotaFormatter(pi.PeeringQuota, "\t\tOld Peering Quota %s"))
+	pi.PeeringQuota = resources.DeepCopy()
+	webhookrefreshlog.Info(quotaFormatter(pi.PeeringQuota, "\t\tNew Peering Quota %s"))
+}
+
+func (pi *peeringInfo) updateFreePeeringQuota(resources v1.ResourceList) {
+	webhookrefreshlog.Info(quotaFormatter(pi.FreePeeringQuota, "\t\tOld Free Quota %s"))
+	pi.FreePeeringQuota = resources.DeepCopy()
+	webhookrefreshlog.Info(quotaFormatter(pi.FreePeeringQuota, "\t\tNew Free Quota %s"))
+}
+
+func (pi *peeringInfo) addShadowPod(spd ShadowPodDescription) {
+	pi.SPList[spd.getName()] = spd
+	pi.subtractResources(spd.Quota)
+}
+
+func (pi *peeringInfo) terminateShadowPod(spd ShadowPodDescription) {
+	spd.terminate()
+	pi.SPList[spd.getName()] = spd
+	pi.addResources(spd.Quota)
+}
+
+func (pi *peeringInfo) removeShadowPod(spd ShadowPodDescription) {
+	delete(pi.SPList, spd.getName())
+}
+
+func (pi *peeringInfo) updateShadowPod(spd ShadowPodDescription) {
+	pi.SPList[spd.getName()] = spd
+}
+
+func (pi *peeringInfo) getQuota() v1.ResourceList {
+	return pi.PeeringQuota
+}
+
+func (pi *peeringInfo) getUsedQuota() v1.ResourceList {
+	UsedQuota := v1.ResourceList{}
+	Quota := pi.getQuota()
+	FreeQuota := pi.getFreeQuota()
+	for key, val := range Quota {
+		tmpQuota := val.DeepCopy()
+		tmpQuota.Sub(FreeQuota[key])
+		UsedQuota[key] = tmpQuota
+	}
+	return UsedQuota
+}
+
+func (pi *peeringInfo) getFreeQuota() v1.ResourceList {
+	return pi.FreePeeringQuota
+}
+
+func (pi *peeringInfo) getShadowPodDescription(uid string) (ShadowPodDescription, bool) {
+	spd, ok := pi.SPList[uid]
+	return spd, ok
+}
+
+func (pi *peeringInfo) getAllShadowPodDescription() map[string]ShadowPodDescription {
+	return pi.SPList
+}
+
+func (pi *peeringInfo) testAndUpdatePeeringInfo(spd ShadowPodDescription, operation admissionv1.Operation, dryRun bool) error {
```

Comment:
Would personally create two separate funcitons for increased clarity, as practically no code is reused.